### PR TITLE
PLANNER-1230: Use Bootstrap based week-picker, date-picker, date-time-picker

### PR DESCRIPTION
**Pull request by Christopher-Chianelli**
Added two dependencies that are required to use the one Patternfly uses:

- MomentJS (https://momentjs.com/), uses MIT license
- BootscrapDateTimePicker (https://eonasdan.github.io/bootstrap-datetimepicker/), uses MIT license
(https://github.com/Eonasdan/bootstrap-datetimepicker/blob/master/LICENSE) (for some reason they didn't put license info on the main website, but this is the date time picker Patternfly uses)

### CODE REVIEW

File: employee-rostering-gwtui/src/main/resources/org/optaweb/employeerostering/gwtui/public/gwtui.html [hunk 1]
```diff
@@ -32,10 +36,16 @@ <h1>Loading..</h1>
 
 <script type="text/javascript">
     erraiBusRemoteCommunicationEnabled = false;
+
+
+
+
```
[the remark this answers is not in the document]
Comment (Christopher-Chianelli) [reply employee-rostering-gwtui/src/main/resources/org/optaweb/employeerostering/gwtui/public/gwtui.html]: Fixed

File: employee-rostering-gwtui/src/main/resources/org/optaweb/employeerostering/gwtui/client/common/LocalDatePicker.html [hunk 2]
```diff
@@ -1,7 +1,6 @@
-<div>
-  <!-- Ideally by the year 9999 all browser support input type date so we don't need pattern -->
-  <!-- TODO: Load pattern from i18n (pattern is ignored by browser, so no way to set
-       date format on browsers that do support input type date) -->
-  <input class="form-control" data-field="date-picker" type="date"
-  pattern="(0[1-9]|1[0-9]|2[0-9]|3[01])/(0[1-9]|1[012])/[0-9]{4}" required></input>
+<div class="input-group date-time-picker-pf" data-field="date-picker">
+    <input type="text" data-field="date-picker-text-field" class="form-control" style="min-width: 250px;">
```
[the remark this answers is not in the document]
Comment (Christopher-Chianelli) [reply employee-rostering-gwtui/src/main/resources/org/optaweb/employeerostering/gwtui/client/common/LocalDatePicker.html]: Fixed

File: employee-rostering-gwtui/src/main/resources/org/optaweb/employeerostering/gwtui/client/common/LocalDatePicker.html [hunk 3]
```diff
@@ -1,7 +1,6 @@
...
+    <span class="input-group-addon">
+                <span class="fa fa-calendar"></span>
+            </span>
```
[the remark this answers is not in the document]
Comment (Christopher-Chianelli) [reply employee-rostering-gwtui/src/main/resources/org/optaweb/employeerostering/gwtui/client/common/LocalDatePicker.html]: Fixed

File: employee-rostering-gwtui/src/main/resources/org/optaweb/employeerostering/gwtui/client/common/LocalDateTimePicker.html [hunk 4]
```diff
@@ -1,9 +1,6 @@
-<div class="local-date-time-picker">
-  <!-- Ideally by the year 9999 all browser support input type date so we don't need pattern -->
-  <!-- TODO: Load pattern from i18n (pattern is ignored by browser, so no way to set
-       date format on browsers that do support input type date) -->
-  <input class="form-control" data-field="date-picker" type="date"
-  pattern="(0[1-9]|1[0-9]|2[0-9]|3[01])/(0[1-9]|1[012])/[0-9]{4}" required></input>
-  <input class="form-control" data-field="time-picker" type="time"
-  pattern="[012]?[0-9]:[012345][0-9]" required></input>
+<div class="input-group date-time-picker-pf" data-field="date-time-picker">
+    <input type="text" data-field="date-time-picker-text-field" class="form-control" style="min-width: 250px;">
```
[the remark this answers is not in the document]
Comment (Christopher-Chianelli) [reply employee-rostering-gwtui/src/main/resources/org/optaweb/employeerostering/gwtui/client/common/LocalDateTimePicker.html]: Fixed

File: employee-rostering-gwtui/src/main/resources/org/optaweb/employeerostering/gwtui/client/common/LocalWeekDatePicker.html [hunk 5]
```diff
@@ -2,20 +2,15 @@
...
-                                        padding-left: 5px;
-                                        padding-right: 5px;
-                                        padding-top: 3px;">
-        </label>
-        <input class="form-control" id="date-picker" type="date"
-               pattern="(0[1-9]|1[0-9]|2[0-9]|3[01])/(0[1-9]|1[012])/[0-9]{4}" required/>
-        <button data-field="today-button" data-i18n-key="General.today"></button>
+    <li>
+        <div class="weekly-date-picker input-group date-time-picker-pf" data-field="weekly-date-picker">
+            <input type="text" data-field="weekly-date-picker-text-field" class="form-control" style="min-width: 250px;">
```
[the remark this answers is not in the document]
Comment (Christopher-Chianelli) [reply employee-rostering-gwtui/src/main/resources/org/optaweb/employeerostering/gwtui/client/common/LocalWeekDatePicker.html]: Fixed

File: employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/LocalDatePicker.java [hunk 6]
```diff
@@ -17,60 +17,89 @@
...
+        BootstrapDateTimePicker.BootstrapDateTimePickerIcons icons = new BootstrapDateTimePicker.BootstrapDateTimePickerIcons();
+        icons.today = "today-button-pf";
+        options.icons = icons;
+        datePicker.datetimepicker(options);
+        datePicker.on("dp.show", () -> {
+            JQuery.get(datePickerElement).children(".bootstrap-datetimepicker-widget").css("height", "min-content").css("margin-bottom", "-100%");
+        });
+        datePicker.on("dp.change", () -> {
+            String value = datePickerTextField.value;
+            if (value != null && !LocalDate.parse(MomentJs.moment(value).format("YYYY-MM-DD")).equals(currentValue)) {
```
[the remark this answers is not in the document]
Comment (Christopher-Chianelli) [reply employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/LocalDatePicker.java]: Changed it to a not null check with a `ValueChangedEvent.fireIfNotEqual()`

File: employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/LocalDateTimePicker.java [hunk 7]
```diff
@@ -16,55 +16,110 @@
...
+    }
 
     @Override
     public void setValue(LocalDateTime value) {
-        datePicker.valueAsDate = new Date(Date.parse(value.toLocalDate().toString()));
-        timePicker.value = value.toLocalTime().toString();
+        if (currentValue == null || !currentValue.equals(value)) {
+            currentValue = value;
+            if (value != null) {
+                ((BootstrapDateTimePicker.BootstrapDateTimePickerData) datePicker.data("DateTimePicker")).date(MomentJs.moment(value.toString(), "YYYY-MM-DDTHH:mm").toDate());
```
[the remark this answers is not in the document]
Comment (Christopher-Chianelli) [reply employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/LocalDateTimePicker.java]: Added some new lines between method calls, which hopefully makes it easier to read.

File: employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/LocalDateTimePicker.java [hunk 8]
```diff
@@ -16,55 +16,110 @@
...
+        BootstrapDateTimePicker.BootstrapDateTimePickerIcons icons = new BootstrapDateTimePicker.BootstrapDateTimePickerIcons();
+        icons.today = "today-button-pf";
+        options.icons = icons;
+        datePicker.datetimepicker(options);
+        datePicker.on("dp.show", () -> {
+            JQuery.get(datePickerElement).children(".bootstrap-datetimepicker-widget").css("height", "min-content").css("margin-bottom", "-100%");
+        });
+        datePicker.on("dp.change", () -> {
+            String value = datePickerTextField.value;
+            if (value != null && !LocalDateTime.parse(MomentJs.moment(value).format("YYYY-MM-DDTHH:mm")).equals(currentValue)) {
```
[the remark this answers is not in the document]
Comment (Christopher-Chianelli) [reply employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/common/LocalDateTimePicker.java]: I changed it to use `moment(String, format)` parsing.

File: employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/tenant/NewTenantForm.java [hunk 9]
```diff
@@ -187,6 +187,6 @@ private static int valueAsInt(HTMLInputElement inputElement) {
     }
 
     private static native String getSystemTimezone() /*-{
-        return Intl.DateTimeFormat().resolvedOptions().timeZone;
-    }-*/;
+                                                     return Intl.DateTimeFormat().resolvedOptions().timeZone;
+                                                     }-*/;
```
Comment (yurloc) [inline employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/tenant/NewTenantForm.java]: This is unexpected. The original format was correct. I recommend to revert this change.

Comment (Christopher-Chianelli) [reply employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/tenant/NewTenantForm.java]: This was caused from the Eclipse formatter formatting the code (since it is a comment and I didn't set the formatter on/off tags yet since my latest Eclipse reboot) and the IDEA formatter **not** reformatting the code. There is a minor issue with the Pull Request, where the Schedule Start Date field in the new Tenant Form is not pre-populated when the form is created, but is when you click on it. None of the other forms have this issue. Also an issue in Firefox where the
"Today" button has no background (not an issue in Chrome). Will fix the formatting in a commit that also fixes those issues.

**Comment by Christopher-Chianelli** [comment]
@yurloc please review the latest changes (fixed setValue() of LocalDatePicker and fixed Today button being invisible in Firefox, as well as the indentation).

**Comment by yurloc** [comment]
@Christopher-Chianelli Is this expected?
employee rostering as a service - mozilla firefox_080
Looks like there is still no background. And it looked the same before your fix. Maybe I didn't understand the issue. I'm using Firefox 62.0.3.

**Comment by Christopher-Chianelli** [comment]
@yurloc yes that is expected. For me, in Firefox Mozilla Firefox 62.0, "Today" had no background, so you just see a hard to see blue link hovering in mid-air. Now it has a white background.
